crates/matrix-sdk-sqlite/src/state_store.rs

```diff
@@ -1531,6 +1551,51 @@ struct ReceiptData {
     user_id: OwnedUserId,
 }
 
+/// Chunk a large query over some keys.
+///
+/// Imagine there is a _dynamic_ query that runs potentially large number of
+/// parameters, so much that the maximum number of parameters can be hit. Then,
+/// this helper is for you. It will execute the query on chunks of parameters.
+async fn chunk_large_query_over<Query, Fut, Res>(
+    mut keys_to_chunk: Vec<Key>,
+    result_capacity: Option<usize>,
+    do_query: Query,
+) -> Result<Vec<Res>>
+where
+    Query: Fn(Vec<Key>) -> Fut,
+    Fut: Future<Output = Result<Vec<Res>, rusqlite::Error>>,
+{
+    let mut all_results = if let Some(capacity) = result_capacity {
+        Vec::with_capacity(capacity)
+    } else {
+        Vec::new()
+    };
+
+    // `Limit` has a `repr(i32)`, it's safe to cast it to `i32`. Then subtract 1 in
+    // case of.
+    let maximum_chunk_size = Limit::SQLITE_LIMIT_VARIABLE_NUMBER as i32 - 1;
```

Comment:
This won't work, will it? If the query contains a handful of non-chunked placeholders already, adding `SQLITE_LIMIT_VARIABLE_NUMBER - 1` placeholders will still overflow. I'd rather do `SQLITE_LIMIT_VARIABLE_NUMBER / 2` or sth. like that.

Reply:
Good catch!!